### PR TITLE
TM accept external predictive input

### PR DESCRIPTION
**Pull request by ctrl-z-9000-times**
Implements issue #88 

In order to make this work I needed to reorder the following two function calls:
```
TM::Compute() {
  activateCells(activeColumnsSize, activeColumns, learn);
  activateDendrites(learn);

^ WAS | IS NOW v

  activateDendrites(learn);
  activateCells(activeColumnsSize, activeColumns, learn);
}
```

I think that they were originally done in this order so that the after every compute cycle, the TM has valid predictions on hand.  This PR needs the calls done in the new order.  This breaks several unit tests, we will need to think about how to incorporate this PR without breaking API too badly.

### CODE REVIEW

**Comment by breznak** [comment]
> This breaks several unit tests, we will need to think about how to incorporate this PR without breaking API too badly. 

Unit tests probably check the internal-workings, so we'll have to adapt them to the new change. 
But API should stay the same at the end of the compute(), regardless how the insides work...(?)

**Comment by breznak** [comment]
@ctrl-z-9000-times PS: next time please open a branch in the community repo here, instead of yours. And then open a PR from the feature -> master. The advantage is developers (us) can push to that branch directly, when there are some small fixes in the PR, or we want to help...

**Comment by dkeeney** [comment]
But be careful when you push a branch to origin.  On the github page it will ask you which master to attach it to.  The default is the numenta repository not the htm-community repository so you have to change it (at least I do anyway).  Twice I forgot to do that and the folks at numenta were wondering what this PR was for.  Very embarrassing.

File: bindings/py/cpp_src/bindings/algorithms/py_HTM.cpp [hunk 1]
```diff
@@ -127,7 +127,7 @@ namespace nupic_ext
             return py::array_t<nupic::UInt32>(activeCells.size(), activeCells.data());
         });
 
-        py_HTM.def("getPredictiveCells", [](const HTM_t& self)
+        py_HTM.def("getPredictiveCells", [](HTM_t& self)
```
Comment (breznak) [inline bindings/py/cpp_src/bindings/algorithms/py_HTM.cpp]: I'm not a fan of getters not being const (like the stuff with SDRs), why is this needed?

File: src/nupic/algorithms/TemporalMemory.cpp [hunk 2]
```diff
@@ -140,10 +140,7 @@ void TemporalMemory::initialize(
   maxSynapsesPerSegment_ = maxSynapsesPerSegment;
   iteration_ = 0;
 
-  activeCells_.clear();
-  winnerCells_.clear();
-  activeSegments_.clear();
-  matchingSegments_.clear();
+  reset();
```
Comment (breznak) [inline src/nupic/algorithms/TemporalMemory.cpp]: :+1:

File: src/nupic/algorithms/TemporalMemory.cpp [hunk 3]
```diff
@@ -452,6 +449,11 @@ void TemporalMemory::activateCells(size_t activeColumnsSize,
         << "The activeColumns must be a sorted list of indices without "
            "duplicates.";
   }
+  if (learn) {
+    for (Segment segment : activeSegments_) {
```
Comment (breznak) [inline src/nupic/algorithms/TemporalMemory.cpp]: const &

File: src/nupic/algorithms/TemporalMemory.cpp [hunk 4]
```diff
@@ -508,9 +510,23 @@ void TemporalMemory::activateCells(size_t activeColumnsSize,
...
 }
 
-void TemporalMemory::activateDendrites(bool learn) {
+void TemporalMemory::activateDendrites(bool learn,
+                                       const vector<UInt> &extraActive,
+                                       const vector<UInt> &extraWinners) {
+
+  for(const auto &active : extraActive) {
+    NTA_ASSERT( active < extra_ );
+    activeCells_.push_back( active + numberOfCells() );
```
Comment (breznak) [inline src/nupic/algorithms/TemporalMemory.cpp]: add somewhere doc like: "Ranges for cell indices: |0..<= normal cells activation < numberOfCells | <= xxx activation <  extraActive| <= winners ... < extraWinners|" 
...or something along this meaning

File: src/nupic/algorithms/TemporalMemory.cpp [hunk 5]
```diff
@@ -544,38 +560,27 @@ void TemporalMemory::activateDendrites(bool learn) {
...
 
-  for(const auto &active : extraActive) {
-    NTA_ASSERT( active < extra_ );
-    activeCells_.push_back( active + numberOfCells() );
-  }
-  for(const auto &winner : extraWinners) {
-    NTA_ASSERT( winner < extra_ );
-    winnerCells_.push_back( winner + numberOfCells() );
-  }
+  if( !segmentsValid_ )
```
Comment (breznak) [inline src/nupic/algorithms/TemporalMemory.cpp]: use brackets {}

File: src/nupic/algorithms/TemporalMemory.cpp [hunk 6]
```diff
@@ -598,7 +614,13 @@ UInt TemporalMemory::numberOfCells(void) const { return connections.numCells();
...
+vector<CellIdx> TemporalMemory::getPredictiveCells(
+      const vector<UInt> &extraActive,
+      const vector<UInt> &extraWinners) {
+
+  if( !segmentsValid_ )
+    activateDendrites(false, extraActive, extraWinners);
+
   vector<CellIdx> predictiveCells;
 
   for (auto segment = activeSegments_.begin(); segment != activeSegments_.end();
```
Comment (breznak) [inline src/nupic/algorithms/TemporalMemory.cpp]: nit, cbegin(), cend

File: bindings/py/cpp_src/bindings/algorithms/py_HTM.cpp (hunk 1 above)
Comment (breznak) [reply bindings/py/cpp_src/bindings/algorithms/py_HTM.cpp]: > When the active cells changes, the flag is cleared. When the active/matching segments are needed they are computed, and the flag is set. 

so could the getters stay as is (const) and add a new version updatePredictiveCells which would call activateDendrites? 

That way get* is fast, and you add your functionality (even w/o API change)

Comment (ctrl-z-9000-times) [reply bindings/py/cpp_src/bindings/algorithms/py_HTM.cpp]: If the getters are constant and they encounter `segmentsValid_ == false` then they will by necessity raise an exception.  The error message will explain that the user needs to call the method `updatePredictiveCells` before using the getters.

File: src/nupic/algorithms/TemporalMemory.hpp [hunk 7]
```diff
@@ -463,6 +465,7 @@ namespace temporal_memory {
   Permanence permanenceIncrement_;
   Permanence permanenceDecrement_;
   Permanence predictedSegmentDecrement_;
+  UInt extra_;
```
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/TemporalMemory.hpp]: It can't be constant since initialize() needs to set it.
Its a private member so it should be safe regardless.

**Comment by ctrl-z-9000-times** [comment]
Ok @breznak, I made the changes which you requested.  I think that this is ready for another round of review.

File: src/nupic/algorithms/TemporalMemory.cpp [hunk 8]
```diff
@@ -509,9 +505,28 @@ void TemporalMemory::activateCells(size_t activeColumnsSize,
...
+  segmentsValid_ = false;
 }
 
-void TemporalMemory::activateDendrites(bool learn) {
+void TemporalMemory::activateDendrites(bool learn,
+                                       const vector<UInt> &extraActive,
+                                       const vector<UInt> &extraWinners)
+{
+  if( segmentsValid_ ) {
+    NTA_CHECK( extraActive.empty() )  << "tm.activateDendrites() already called";
```
Comment (breznak) [inline src/nupic/algorithms/TemporalMemory.cpp]: should we throw, or just NTA_INFO && return?

File: src/nupic/algorithms/TemporalMemory.hpp [hunk 9]
```diff
@@ -198,8 +215,19 @@ namespace temporal_memory {
...
    * used during segment cleanup.
+   *
+   * @param extraActive
+   * Vector of active external predictive inputs.  External inputs must be cell
+   * indexes in the range [0, extra).
+   *
+   * @param extraWinners
+   * Vector of winning external predictive inputs.  When learning, only these
+   * inputs are considered active.  External inputs must be cell indexes in the
+   * range [0, extra).
```
Comment (breznak) [inline src/nupic/algorithms/TemporalMemory.hpp]: isn't that extWinners is a subset of extActive?

File: src/test/unit/algorithms/TemporalMemoryTest.cpp [hunk 10]
```diff
@@ -1520,6 +1529,72 @@ TEST(TemporalMemoryTest, testSaveLoad) {
...
+      for(UInt i = 0; i < predictedColumns.size(); i++) {
+        predictedColumns[i] /= tm.getCellsPerColumn();
+        if(i > 0 && predictedColumns[i] == predictedColumns[i-1])
+          predictedColumns.erase( predictedColumns.begin() + i-- );
+      }
+      // Calculate TM output
+      const auto &sparse = x.getFlatSparse();
+      tm.compute(sparse.size(), sparse.data(), true);
+      // Testing the test: commenting out the next two lines should cause this
+      // test to fail.
```
Comment (breznak) [inline src/test/unit/algorithms/TemporalMemoryTest.cpp]: please add a check for that

> tm.activateDendrites(true, {}, {}); 

auto predictedColumnsWithoutExtra = tm.getPredictiveCells();

vs

> tm.activateDendrites(true, extraActive, extraWinners); 

auto predictedColumns = tm.getPredictiveCells();
should show the difference in passing the threshold

File: src/nupic/algorithms/TemporalMemory.cpp (hunk 8 above)
Comment (breznak) [reply src/nupic/algorithms/TemporalMemory.cpp]: I see. So the correct coding-pattern is: activate(), compute(), get() ...maybe mention that?

File: src/nupic/algorithms/TemporalMemory.hpp (hunk 9 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/TemporalMemory.hpp]: I will add this to the documentation.

**Comment by ctrl-z-9000-times** [comment]
Okay @breznak ,
* I added to the docs that extraWinners is supposed to be a subset of extraActive.  This is not verified in debug mode. 
    - I'm more worried that the user will forget to give this argument altogether and it will use the default argument (empty list) but this condition can't be checked for.
* I improved the unit test as you requested.

File: src/nupic/algorithms/TemporalMemory.hpp [hunk 11]
```diff
@@ -463,9 +503,11 @@ namespace temporal_memory {
   Permanence permanenceIncrement_;
   Permanence permanenceDecrement_;
   Permanence predictedSegmentDecrement_;
+  UInt extra_;
```
Comment (breznak) [inline src/nupic/algorithms/TemporalMemory.hpp]: serialization & equals have to change for extra_, and possibly also segmentsValid_

**Comment by ctrl-z-9000-times** [comment]
> I'm more worried that the user will forget to give this argument altogether and it will use the default argument (empty list) but this condition can't be checked for.

I found a way to check for this:
```
virtual void compute(size_t activeColumnsSize, const UInt activeColumns[],
    bool learn = true,
    const vector<UInt> &extraActive  = {std::numeric_limits<UInt>::max()},
    const vector<UInt> &extraWinners = {std::numeric_limits<UInt>::max()});
```
And:
```
if( extraActive.size() == 1 && extraActive[0] == std::numeric_limits<UInt>::max() )
```

Now I can make this assert that the correct arguments are given every time.

**Comment by ctrl-z-9000-times** [comment]
Okay @breznak:  I addressed your concerns and this PR is ready for another review when you get the time.

**Comment by ctrl-z-9000-times** [comment]
Hi @breznak ,
Now that you're back can you take a second look at this PR?
Thanks!

**Comment by ctrl-z-9000-times** [comment]
> @ctrl-z-9000-times consider mentioning the new change in API changelog? (does not break anything, but can be used differently now)

Maybe we should add a new section to the API_CHANGELOG.md for new features / functionality?

**Comment by breznak** [comment]
> Maybe we should add a new section to the  

Same would be the new #209
